File: shared/src/main/java/xyz/klinker/messenger/shared/util/NotificationUtils.java
package xyz.klinker.messenger.shared.util;

import android.annotation.TargetApi;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationChannelGroup;
import android.app.NotificationManager;
import android.content.Context;
import android.media.AudioAttributes;
import android.net.Uri;
import android.os.Build;
import android.service.notification.StatusBarNotification;
import android.support.v4.app.NotificationManagerCompat;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import xyz.klinker.messenger.shared.R;
import xyz.klinker.messenger.shared.data.DataSource;
import xyz.klinker.messenger.shared.data.Settings;
import xyz.klinker.messenger.shared.data.model.Conversation;
import xyz.klinker.messenger.shared.service.NotificationService;

public class NotificationUtils {

    public static final String MESSAGE_GROUP_SUMMARY_CHANNEL_ID = "message-group-summary";
    public static final String FAILED_MESSAGES_CHANNEL_ID = "failed-messages";
    public static final String TEST_NOTIFICATIONS_CHANNEL_ID = "test-notifications";
    public static final String STATUS_NOTIFICATIONS_CHANNEL_ID = "status-notifications";
    public static final String MEDIA_PARSE_CHANNEL_ID = "media-parsing";
    public static final String GENERAL_CHANNEL_ID = "general";

    public static void cancelGroupedNotificationWithNoContent(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            Map<String, Integer> map = new HashMap();

            NotificationManager manager = (NotificationManager) context.getSystemService(
                    Context.NOTIFICATION_SERVICE);

            StatusBarNotification[] notifications = manager.getActiveNotifications();

            for (StatusBarNotification notification : notifications) {
                String keyString = notification.getGroupKey();
                if (keyString.contains("|g:")) { // this is a grouped notification
                    keyString = keyString.substring(keyString.indexOf("|g:") + 3, keyString.length());

                    if (map.containsKey(keyString)) {
                        map.put(keyString, map.get(keyString) + 1);
                    } else {
                        map.put(keyString, 1);
                    }
                }
            }

            Iterator it = map.entrySet().iterator();
            while (it.hasNext()) {
                Map.Entry pair = (Map.Entry)it.next();
                String key = (String) pair.getKey();
                int value = (Integer) pair.getValue();

                if (value == 1) {
                    for (StatusBarNotification notification : notifications) {
                        String keyString = notification.getGroupKey();
                        if (keyString.contains("|g:")) { // this is a grouped notification
                            keyString = keyString.substring(keyString.indexOf("|g:") + 3, keyString.length());

                            if (key.equals(keyString)) {
                                manager.cancel(notification.getId());
                                break;
                            }
                        }
                    }
                }

                it.remove();
            }
        }
    }

    @TargetApi(Build.VERSION_CODES.O)
    public static void createNotificationChannelIfNonExistent(Context context, Conversation conversation) {
        if (!AndroidVersionUtil.isAndroidO()) {
            return;
        }

        final NotificationManager manager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        final NotificationChannel channel = createChannel(context, conversation);
        manager.createNotificationChannel(channel);
    }

    @TargetApi(Build.VERSION_CODES.O)
    public static void createTestChannel(Context context) {
        if (!AndroidVersionUtil.isAndroidO()) {
            return;
        }

        final NotificationManager manager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        NotificationChannel testChannel = new NotificationChannel(TEST_NOTIFICATIONS_CHANNEL_ID,
                context.getString(R.string.test_notifications_channel),
                Settings.get(context).headsUp ? NotificationManager.IMPORTANCE_MAX : NotificationManager.IMPORTANCE_DEFAULT);
        manager.createNotificationChannel(testChannel);
    }

    @TargetApi(Build.VERSION_CODES.O)
    public static void createStatusChannel(Context context) {
        if (!AndroidVersionUtil.isAndroidO()) {
            return;
        }

        final NotificationManager manager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        NotificationChannel statusChannel = new NotificationChannel(STATUS_NOTIFICATIONS_CHANNEL_ID,
                context.getString(R.string.status_notifications_channel), NotificationManager.IMPORTANCE_DEFAULT);
        manager.createNotificationChannel(statusChannel);
    }

    @TargetApi(Build.VERSION_CODES.O)
    public static void createMediaParseChannel(Context context) {
        if (!AndroidVersionUtil.isAndroidO()) {
            return;
        }

        final NotificationManager manager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        NotificationChannel statusChannel = new NotificationChannel(MEDIA_PARSE_CHANNEL_ID,
                context.getString(R.string.status_notifications_channel), NotificationManager.IMPORTANCE_MIN);
        manager.createNotificationChannel(statusChannel);
    }

    @TargetApi(Build.VERSION_CODES.O)
    public static void createGeneralChannel(Context context) {
        if (!AndroidVersionUtil.isAndroidO()) {
            return;
        }

        final NotificationManager manager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        NotificationChannel statusChannel = new NotificationChannel(GENERAL_CHANNEL_ID,
                context.getString(R.string.general_notifications_channel), NotificationManager.IMPORTANCE_MIN);
        manager.createNotificationChannel(statusChannel);
    }

    @TargetApi(Build.VERSION_CODES.O)
    public static void createFailedMessageChannel(Context context) {
        if (!AndroidVersionUtil.isAndroidO()) {
            return;
        }

        final NotificationManager manager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        NotificationChannel testChannel = new NotificationChannel(FAILED_MESSAGES_CHANNEL_ID,
                context.getString(R.string.failed_messages_channel),
                Settings.get(context).headsUp ? NotificationManager.IMPORTANCE_MAX : NotificationManager.IMPORTANCE_DEFAULT);
        manager.createNotificationChannel(testChannel);
    }

    @TargetApi(Build.VERSION_CODES.O)
    public static void createMessageGroupChannel(Context context) {
        if (!AndroidVersionUtil.isAndroidO()) {
            return;
        }

        final NotificationManager manager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        final NotificationChannel messageGroupChannel = new NotificationChannel(MESSAGE_GROUP_SUMMARY_CHANNEL_ID,
                context.getString(R.string.group_summary_notifications),
                Settings.get(context).headsUp ? NotificationManager.IMPORTANCE_MAX : NotificationManager.IMPORTANCE_DEFAULT);
        manager.createNotificationChannel(messageGroupChannel);
    }

    @TargetApi(Build.VERSION_CODES.O)
    public static void createNotificationChannels(Context context, DataSource source) {
        if (!AndroidVersionUtil.isAndroidO()) {
            return;
        }

        final NotificationManager manager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

        // notification channel group for conversations
        NotificationChannelGroup conversationsGroup = new NotificationChannelGroup("conversations",
                context.getString(R.string.conversations));
        manager.createNotificationChannelGroup(conversationsGroup);

        // channels to place the notifications in
        createTestChannel(context);
        createStatusChannel(context);
        createFailedMessageChannel(context);
        createMessageGroupChannel(context);
        createMediaParseChannel(context);
        createGeneralChannel(context);

        List<Conversation> conversations = source.getAllConversationsAsList();
        for (int i = 0; i < conversations.size(); i++) {
            final Conversation conversation = conversations.get(i);
            final NotificationChannel channel = createChannel(context, conversation);

            try {
                manager.createNotificationChannel(channel);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    @TargetApi(Build.VERSION_CODES.O)
    private static NotificationChannel createChannel(Context context, Conversation conversation) {
        Settings settings = Settings.get(context);

        NotificationChannel channel = new NotificationChannel(conversation.id + "", conversation.title,
                settings.headsUp ? NotificationManager.IMPORTANCE_MAX : NotificationManager.IMPORTANCE_DEFAULT);
        channel.setGroup("conversations");
        channel.enableLights(true);
        channel.setLightColor(conversation.ledColor);
        channel.setBypassDnd(false);
        channel.setShowBadge(true);
        channel.setVibrationPattern(Settings.get(context).vibrate.pattern);
        channel.setLockscreenVisibility(conversation.privateNotifications ?
                Notification.VISIBILITY_PRIVATE : Notification.VISIBILITY_PUBLIC);

        Uri ringtone = NotificationService.getRingtone(context, conversation.ringtoneUri);
        if (ringtone != null) {
            channel.setSound(ringtone, new AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_NOTIFICATION).build());
        }

        return channel;
    }

    @TargetApi(Build.VERSION_CODES.O)
    public static void deleteChannel(Context context, long conversationId) {
        if (!AndroidVersionUtil.isAndroidO()) {
            return;
        }

        final NotificationManager manager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        manager.deleteNotificationChannel(conversationId + "");
    }

}

File: shared/src/main/java/xyz/klinker/messenger/shared/util/NotificationUtils.kt
package xyz.klinker.messenger.shared.util

import android.annotation.TargetApi
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationChannelGroup
import android.app.NotificationManager
import android.content.Context
import android.graphics.Color
import android.os.Build
import xyz.klinker.messenger.shared.R
import xyz.klinker.messenger.shared.data.DataSource
import xyz.klinker.messenger.shared.data.Settings
import xyz.klinker.messenger.shared.data.model.Conversation

object NotificationUtils {

    const val DEFAULT_CONVERSATION_CHANNEL_ID = "default-conversation-channel"
    const val SILENT_CONVERSATION_CHANNEL_ID = "silent-conversation-channel"
    const val QUICK_TEXT_CHANNEL_ID = "quick-text"
    const val SILENT_BACKGROUND_CHANNEL_ID = "silent-background-services"
    const val ACCOUNT_ACTIVITY_CHANNEL_ID = "account-activity-channel"

    fun cancelGroupedNotificationWithNoContent(context: Context?) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && context != null) {
            Thread {
                try {
                    Thread.sleep(100)
                } catch (e: Exception) {
                }

                val cursor = DataSource.getUnseenMessages(context)
                if (cursor.count == 0) {
                    // all messages are seen
                    val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

                    try {
                        manager.cancelAll()
                    } catch (e: SecurityException) {
                        // not the right permissions. Changed user account maybe
                    }
                }

                cursor.closeSilent()
            }.start()
        }
    }

    @TargetApi(Build.VERSION_CODES.O)
    fun createNotificationChannel(context: Context, conversation: Conversation) {
        if (!AndroidVersionUtil.isAndroidO) {
            return
        }

        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val channel = createChannel(context, conversation)
        manager.createNotificationChannel(channel)
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun createDefaultChannel(context: Context) {
        if (!AndroidVersionUtil.isAndroidO) {
            return
        }

        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val defaultChannel = NotificationChannel(DEFAULT_CONVERSATION_CHANNEL_ID,
                context.getString(R.string.default_notifications_channel), NotificationManager.IMPORTANCE_HIGH)
        defaultChannel.description = context.getString(R.string.default_notifications_channel_description)
        defaultChannel.group = "conversations"
        defaultChannel.lightColor = Color.WHITE
        defaultChannel.enableLights(true)
        defaultChannel.setBypassDnd(false)
        defaultChannel.setShowBadge(true)
        defaultChannel.enableVibration(true)
        manager.createNotificationChannel(defaultChannel)
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun createQuickTextChannel(context: Context) {
        if (!AndroidVersionUtil.isAndroidO) {
            return
        }

        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val testChannel = NotificationChannel(QUICK_TEXT_CHANNEL_ID,
                context.getString(R.string.quick_text_channel), NotificationManager.IMPORTANCE_MIN)
        testChannel.setShowBadge(false)
        testChannel.enableLights(false)
        testChannel.enableVibration(false)
        manager.createNotificationChannel(testChannel)
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun createSilentConversationChannel(context: Context) {
        if (!AndroidVersionUtil.isAndroidO) {
            return
        }

        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val silentConversations = NotificationChannel(SILENT_CONVERSATION_CHANNEL_ID,
                context.getString(R.string.silent_conversations_channel), NotificationManager.IMPORTANCE_LOW)
        silentConversations.group = "conversations"
        silentConversations.setShowBadge(true)
        silentConversations.enableLights(true)
        silentConversations.enableVibration(false)
        manager.createNotificationChannel(silentConversations)
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun createSilentBackgroundChannel(context: Context) {
        if (!AndroidVersionUtil.isAndroidO) {
            return
        }

        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val silentBackground = NotificationChannel(SILENT_BACKGROUND_CHANNEL_ID,
                context.getString(R.string.silent_background_services), NotificationManager.IMPORTANCE_MIN)
        silentBackground.setShowBadge(false)
        silentBackground.enableLights(false)
        silentBackground.enableVibration(false)
        manager.createNotificationChannel(silentBackground)
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun createAccountActivityChannel(context: Context) {
        if (!AndroidVersionUtil.isAndroidO) {
            return
        }

        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val silentBackground = NotificationChannel(ACCOUNT_ACTIVITY_CHANNEL_ID,
                context.getString(R.string.account_activity_notifications), NotificationManager.IMPORTANCE_LOW)
        silentBackground.setShowBadge(false)
        silentBackground.enableLights(false)
        silentBackground.enableVibration(false)
        manager.createNotificationChannel(silentBackground)
    }

    @TargetApi(Build.VERSION_CODES.O)
    fun createNotificationChannels(context: Context) {
        if (!AndroidVersionUtil.isAndroidO) {
            return
        }

        deleteOldChannels(context)

        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        // notification channel group for conversations
        val conversationsGroup = NotificationChannelGroup("conversations", context.getString(R.string.conversations))
        manager.createNotificationChannelGroup(conversationsGroup)

        // channels to place the notifications in
        createDefaultChannel(context)
        createQuickTextChannel(context)
        createSilentConversationChannel(context)
        createSilentBackgroundChannel(context)
        createAccountActivityChannel(context)
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun createChannel(context: Context, conversation: Conversation): NotificationChannel {
        val settings = Settings

        val channel = NotificationChannel(conversation.id.toString() + "", conversation.title, NotificationManager.IMPORTANCE_HIGH)
        channel.group = "conversations"
        channel.enableLights(true)
        channel.lightColor = Color.WHITE
        channel.setBypassDnd(false)
        channel.setShowBadge(true)
        channel.enableVibration(true)
        channel.lockscreenVisibility = if (conversation.private)
            Notification.VISIBILITY_PRIVATE else Notification.VISIBILITY_PUBLIC

        return channel
    }

    @TargetApi(Build.VERSION_CODES.O)
    fun deleteChannel(context: Context, conversationId: Long) {
        if (!AndroidVersionUtil.isAndroidO) {
            return
        }

        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        manager.deleteNotificationChannel(conversationId.toString() + "")
    }

    @TargetApi(Build.VERSION_CODES.O)
    fun deleteOldChannels(context: Context) {
        if (!AndroidVersionUtil.isAndroidO) {
            return
        }

        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        manager.deleteNotificationChannel("background-service")
        manager.deleteNotificationChannel("general")
        manager.deleteNotificationChannel("media-parsing")
        manager.deleteNotificationChannel("status-notifications")
        manager.deleteNotificationChannel("test-notifications")
        manager.deleteNotificationChannel("failed-messages")
        manager.deleteNotificationChannel("message-group-summary")
    }

    @TargetApi(Build.VERSION_CODES.O)
    fun deleteAllChannels(context: Context) {
        if (!AndroidVersionUtil.isAndroidO) {
            return
        }

        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val channels = manager.notificationChannels
        for (channel in channels) {
            manager.deleteNotificationChannel(channel.id)
        }
    }

}
